### ISSUE TITLE
Service crashes due to unhandled outbox/NHibernate exception

### ISSUE THREAD

**Issue by macdonald-keith-vmware** [Particular/NServiceBus#4293]
I am using NSB 6.0, RabbitMQ, NHibernate persistence, SQL Server, and the Outbox feature.

We were performing server patches/updates this weekend. During that time period, the SQL Server database server used for NSB persistence was rebooted. My NSB service crashed due to an unhandled exception when NSB was attempting to remove old outbox entries.

call stack follows.
==========================================================

Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.Data.SqlClient.SqlException
   at System.Data.SqlClient.SqlConnection.OnError(System.Data.SqlClient.SqlException, Boolean, System.Action`1<System.Action>)
   at System.Data.SqlClient.SqlInternalConnection.OnError(System.Data.SqlClient.SqlException, Boolean, System.Action`1<System.Action>)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(System.Data.SqlClient.TdsParserStateObject, Boolean, Boolean)
   at System.Data.SqlClient.TdsParser.TryRun(System.Data.SqlClient.RunBehavior, System.Data.SqlClient.SqlCommand, System.Data.SqlClient.SqlDataReader, System.Data.SqlClient.BulkCopySimpleResultSet, System.Data.SqlClient.TdsParserStateObject, Boolean ByRef)
   at System.Data.SqlClient.TdsParser.Run(System.Data.SqlClient.RunBehavior, System.Data.SqlClient.SqlCommand, System.Data.SqlClient.SqlDataReader, System.Data.SqlClient.BulkCopySimpleResultSet, System.Data.SqlClient.TdsParserStateObject)
   at System.Data.SqlClient.TdsParser.TdsExecuteTransactionManagerRequest(Byte[], TransactionManagerRequestType, System.String, TransactionManagerIsolationLevel, Int32, System.Data.SqlClient.SqlInternalTransaction, System.Data.SqlClient.TdsParserStateObject, Boolean)
   at System.Data.SqlClient.SqlInternalConnectionTds.ExecuteTransactionYukon(TransactionRequest, System.String, System.Data.IsolationLevel, System.Data.SqlClient.SqlInternalTransaction, Boolean)
   at System.Data.SqlClient.SqlInternalConnectionTds.ExecuteTransaction(TransactionRequest, System.String, System.Data.IsolationLevel, System.Data.SqlClient.SqlInternalTransaction, Boolean)
   at System.Data.SqlClient.SqlInternalConnection.BeginSqlTransaction(System.Data.IsolationLevel, System.String, Boolean)
   at System.Data.SqlClient.SqlConnection.BeginTransaction(System.Data.IsolationLevel, System.String)
   at System.Data.SqlClient.SqlConnection.BeginDbTransaction(System.Data.IsolationLevel)
   at System.Data.Common.DbConnection.System.Data.IDbConnection.BeginTransaction(System.Data.IsolationLevel)
   at NHibernate.Transaction.AdoTransaction.Begin(System.Data.IsolationLevel)

Exception Info: NHibernate.TransactionException
   at NHibernate.Transaction.AdoTransaction.Begin(System.Data.IsolationLevel)
   at NHibernate.Impl.StatelessSessionImpl.BeginTransaction(System.Data.IsolationLevel)
   at NServiceBus.Persistence.NHibernate.OutboxPersister.RemoveEntriesOlderThan(System.DateTime)
   at NServiceBus.Features.OutboxCleaner.PerformCleanup(System.Object)
   at System.Threading.TimerQueueTimer.CallCallbackInContext(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.TimerQueueTimer.CallCallback()
   at System.Threading.TimerQueueTimer.Fire()
   at System.Threading.TimerQueue.FireQueuedTimerCompletion(System.Object)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()

**Comment by dbelcham** [comment]
Based on what you wrote here it looks like the `OutboxCleaner.PerformCleanup` routine was trying to run when the SQL Server was offline. By default this routine runs every minute. This failure will not cause loss of messages and will attempt to run again when the next scheduled time occurs. If SQL Server is available when that subsequent run occurs no exception will be raised.

To avoid this in the future, you would be best served by stopping the endpoints before stopping SQL Server.

Does this provide you with the information you need/want?

**Comment by macdonald-keith-vmware** [comment]
Let me review the application logs again, because it appeared to me based on what I saw that this exception caused the application to terminate, which I would not expect. I would instead expect the behavior that you described, where it would just retry again in the next interval.

**Comment by ramonsmits** [comment]
@macdonald-keith-vmware I want to advise to you to enable windows service recovery for all your nservicebus endpoints

- https://docs.particular.net/nservicebus/hosting/windows-service#installation-restart-recovery

This will make sure that at least your services will be restarted when this happens.

**Comment by timbussmann** [comment]
I think a failure in `OutboxCleaner.PerformCleanup` should not immediately take down the endpoint and that behavior might be unintended. I'll raise an issue in the NHibernate repository for this.

But note that other components (and even the outbox) might shut down the endpoint if the fail to connect to the sql server after a certain period of time. E.g. the timeout manager will by default shut down the endpoint after 2 minutes but you can configure this value as documented here: https://docs.particular.net/nservicebus/recoverability/critical-exception-for-timeout-outages

So to summarize:
Imho
* The endpoint shut not immediately fail when the outbox cleanup task fails
* The outbox cleanup task might want to use a custom circuit breaker similar to the timeout manager
* You still need to be aware that endpoints might shut down when taking the SQL server down
  * You can configure the circuit breakers to be more patient until the SQL server is back up again
  * You can configure windows service recovery to automatically restart services as described by @ramonsmits

**Comment by timbussmann** [comment]
as https://github.com/Particular/NServiceBus.NHibernate/issues/241 has been confirmed by the nhibernate maintainers, I'm going to close this issue, as there are no direct changes required in the core. 

@macdonald-keith-vmware thanks a lot for raising this issue. With https://github.com/Particular/NServiceBus.NHibernate/issues/241 the exception you received should be fixed. But please not, that as mentioned in previous comments, taking down the SQL server for to long may still result in endpoints shutting down.
